#HomeWork 1
### Submitted by: <b>Soumi Basu (EID: sb59982)</b>



###<b>Question 1</b>
A bank makes four kinds of loans to its customers and these loans yield the following annual interest rates to the bank:

- First mortgage 14%
- Second mortgage 20%
- Home improvement 20%
- Personal overdraft 10%

We are interested in the bank’s lending strategy. The information we know is as following:

1. In total $250 million is lent out.
2. First mortgages are 55% of all mortgages (i.e. first and second mortgage) issued.

3. Second mortgages are 25% of all loans issued.
4. The dollar-weighted average interest rate on all loans is 15%.

Calculate the lending strategy using matrix inversion.  How much is lent in home improvement loans?

 

Answer in millions of dollars, rounded to 2 decimal places.  If the answer is $23080444.12, then you should enter 23.08.

#Answer:
Let the money lent by the bank under First mortgage be A, under Second mortgage be B, Home improvement be C and Personal overdraft be D.

Then from the equations given:

    A + B + C + D = 250 (million)
    A = 0.55(A + B)
    B = 0.25(A + B + C + D)
    0.14A + 0.2B + 0.2C + 0.1D = 0.15*250 = 37.5 (million)



In [2]:
import numpy as np

In [ ]:
A = np.zeros((4,4))
A[0,:] = [1,1,1,1]
A[1,:] = [0.45, -.55,0,0]
A[2,:] = [0.25, -0.75, 0.25, 0.25]
A[3,:] = [0.14,0.2,0.2,0.1]
A

array([[ 1.  ,  1.  ,  1.  ,  1.  ],
       [ 0.45, -0.55,  0.  ,  0.  ],
       [ 0.25, -0.75,  0.25,  0.25],
       [ 0.14,  0.2 ,  0.2 ,  0.1 ]])

In [ ]:
B = np.array([250,0,0,37.5])
B

array([250. ,   0. ,   0. ,  37.5])

In [ ]:
x = np.linalg.solve(A,B)
x

array([76.38888889, 62.5       , 31.94444444, 79.16666667])

The loan amount lend via home improvement loans is C which is **$31.94 million**.




###<b>Question 2</b>


####Part 1

In [10]:
X = np.zeros((10,5))

X[:,0] = [1,1,1,1,0,0,0,0,0,0]
X[:,1] = [-1,0,0,0,1,1,1,0,0,0]
X[:,2] = [0,-1,0,0,-1,0,0,1,1,0]
X[:,3] = [0,0,-1,0,0,-1,0,-1,0,1]
X[:,4] = [0,0,0,-1,0,0,-1,0,-1,-1]

X

array([[ 1., -1.,  0.,  0.,  0.],
       [ 1.,  0., -1.,  0.,  0.],
       [ 1.,  0.,  0., -1.,  0.],
       [ 1.,  0.,  0.,  0., -1.],
       [ 0.,  1., -1.,  0.,  0.],
       [ 0.,  1.,  0., -1.,  0.],
       [ 0.,  1.,  0.,  0., -1.],
       [ 0.,  0.,  1., -1.,  0.],
       [ 0.,  0.,  1.,  0., -1.],
       [ 0.,  0.,  0.,  1., -1.]])

In [23]:
y = np.array([[-45,-3,-31,-45,18,8,20,2,-27,-38]])
Y = y.T
Y

array([[-45],
       [ -3],
       [-31],
       [-45],
       [ 18],
       [  8],
       [ 20],
       [  2],
       [-27],
       [-38]])

####Part 2

In [25]:
M = X.T@X
M

array([[ 4., -1., -1., -1., -1.],
       [-1.,  4., -1., -1., -1.],
       [-1., -1.,  4., -1., -1.],
       [-1., -1., -1.,  4., -1.],
       [-1., -1., -1., -1.,  4.]])

In [26]:
p = X.T@Y
p

array([[-124.],
       [  91.],
       [ -40.],
       [ -17.],
       [  90.]])

####Part 3

In [27]:
M[4,:] = [1,1,1,1,1]
M

array([[ 4., -1., -1., -1., -1.],
       [-1.,  4., -1., -1., -1.],
       [-1., -1.,  4., -1., -1.],
       [-1., -1., -1.,  4., -1.],
       [ 1.,  1.,  1.,  1.,  1.]])

In [28]:
p[4,:] = [0]
p

array([[-124.],
       [  91.],
       [ -40.],
       [ -17.],
       [   0.]])

In [29]:
r = np.linalg.solve(M,p)
r

array([[-24.8],
       [ 18.2],
       [ -8. ],
       [ -3.4],
       [ 18. ]])

**Ratings**

 - Team 2 has the highest rating with a rating of 18.2.
 - **Team 5** has the second highest rating with a rating of 18.
 - Team 4 has the third highest rating with a rating of -3.4.
 - Team 3 has the fourth highest rating with a rating of -8.
 - Team 1 has the lowest rating with a rating of -24.8.

###<b>Question 3</b>

A Lehmer matrix is one whose entries are specified by the following rule:

    Ai,j = i/j if j > i and Ai,j = j/i otherwise

Write a function named lehmer_entry which takes two arguments and outputs the value of the entry. Then use “for loop(s)” to generate a 20 by 20 Lehmer Matrix.

(Hint: First generate a 4 by 4 matrix with all the elements being 0. Then use for loop(s) and if statement to define the Lehmer matrix. Find the 4 by 4 Lehmer matrix and use the Wikipedia to check. Then you can change the code to run a 20 by 20)

Is A symmetric?

In [ ]:
def lehmer_entry(a,b):
  Lehmer = np.zeros((a,b))
  for i in range(a):
    for j in range(b):
      Lehmer[i,j] = float(i+1)/float(j+1) if j>i else float(j+1)/float(i+1)
  return Lehmer
lehmer_entry(4,4)


array([[1.        , 0.5       , 0.33333333, 0.25      ],
       [0.5       , 1.        , 0.66666667, 0.5       ],
       [0.33333333, 0.66666667, 1.        , 0.75      ],
       [0.25      , 0.5       , 0.75      , 1.        ]])

In [ ]:
A = lehmer_entry(20,20)
A

array([[1.        , 0.5       , 0.33333333, 0.25      , 0.2       ,
        0.16666667, 0.14285714, 0.125     , 0.11111111, 0.1       ,
        0.09090909, 0.08333333, 0.07692308, 0.07142857, 0.06666667,
        0.0625    , 0.05882353, 0.05555556, 0.05263158, 0.05      ],
       [0.5       , 1.        , 0.66666667, 0.5       , 0.4       ,
        0.33333333, 0.28571429, 0.25      , 0.22222222, 0.2       ,
        0.18181818, 0.16666667, 0.15384615, 0.14285714, 0.13333333,
        0.125     , 0.11764706, 0.11111111, 0.10526316, 0.1       ],
       [0.33333333, 0.66666667, 1.        , 0.75      , 0.6       ,
        0.5       , 0.42857143, 0.375     , 0.33333333, 0.3       ,
        0.27272727, 0.25      , 0.23076923, 0.21428571, 0.2       ,
        0.1875    , 0.17647059, 0.16666667, 0.15789474, 0.15      ],
       [0.25      , 0.5       , 0.75      , 1.        , 0.8       ,
        0.66666667, 0.57142857, 0.5       , 0.44444444, 0.4       ,
        0.36363636, 0.33333333, 0.30769231, 0

From the above matrix, we can see that matrix A (Lehmer matrix) is **symmetric**.

###Question 4

Going back to the Lehmer matrix problem.

 

    Calculate the inverse of A and assign it to C.
    Assign [1 2 3 4 5 6 7 8 9 10 10 9 8 7 6 5 4 3 2 1] to d.
    Solve for x in the equation Ax = Cd
    What is x10 ? Round to 1 decimal place.  Be careful if the answer you get is in scientific notation.

In [ ]:
C = np.linalg.inv(A)
C

array([[ 1.33333333e+00, -6.66666667e-01,  0.00000000e+00,
        -1.05735526e-16,  7.40148683e-17, -3.70074342e-17,
        -1.11022302e-16, -2.09541178e-31, -6.23283102e-17,
         2.14995570e-16, -2.30089699e-16,  1.11022302e-16,
        -3.70074342e-17, -1.49943914e-17,  1.85037171e-16,
         1.12143740e-18, -1.11022302e-16,  5.55111512e-17,
        -1.00584308e-16,  1.04379942e-17],
       [-6.66666667e-01,  2.13333333e+00, -1.20000000e+00,
        -2.96059473e-16, -9.07915718e-16,  2.96059473e-16,
        -4.14483263e-17,  8.88178420e-17,  1.18813341e-16,
        -3.66549824e-16,  3.15367700e-17,  1.08357767e-16,
         1.18423789e-16, -1.61428980e-16, -1.02188270e-16,
        -4.58443609e-16,  9.76996262e-16, -6.21724894e-16,
         2.87139733e-16, -1.34745017e-16],
       [ 3.30631075e-33, -1.20000000e+00,  3.08571429e+00,
        -1.71428571e+00,  4.73695157e-16, -6.47833435e-16,
         4.42820384e-16, -1.27629000e-16, -2.61695427e-16,
         1.22351109e-16,  1.1

In [ ]:
D = np.array([1,2,3,4,5,6,7,8,9,10,10,9,8,7,6,5,4,3,2,1]) #assigning D

In [ ]:
C@D #multiplying matrices C*D

array([ 0.00000000e+00, -4.44089210e-15,  3.55271368e-15,  0.00000000e+00,
        3.55271368e-15,  2.13162821e-14, -2.84217094e-14,  2.84217094e-14,
       -2.84217094e-14,  5.23809524e+00,  6.19461698e+00,  8.76521739e-01,
        8.08888889e-01,  7.50957854e-01,  7.00778643e-01,  6.56891496e-01,
        6.18181818e-01,  5.83783784e-01,  5.53014553e-01, -9.23076923e+00])

In [ ]:
#solving for Ax = C*D
x = np.linalg.solve(A,C@D) 
x

array([ 2.36847579e-15, -1.43292785e-14,  2.25917387e-14, -2.91360470e-14,
       -3.45973556e-14,  2.21485798e-13, -3.94494813e-13,  4.98702462e-13,
       -2.48120301e+01,  2.00642393e+01,  3.58137507e+01, -3.00626280e+01,
       -3.73699570e-04, -2.77204399e-04, -2.09968769e-04, -1.61954093e-04,
       -1.26922826e-04, -1.00877938e-04,  9.50593309e+01, -1.00062903e+02])

In [ ]:
round(x[9],1) #finding the 10th element (x10).

20.1

##### **THE END**#####